File: bnovate/bnovate/report/work_order_planning/work_order_planning.py
# Copyright (c) 2023, bNovate, libracore, and contributors
# For license information, please see license.txt
#
# WORK ORDER PLANNING
#####################
#
# Show which open work orders are executable based on inventory levels.
#
###################################################################

from __future__ import unicode_literals
import frappe
from frappe import _
from collections import OrderedDict
from urllib.parse import quote

from bnovate.bnovate.report.projected_stock import projected_stock

def execute(filters=None):
    columns = get_columns(filters)
    data = get_data(filters)
    return columns, data

def get_columns(filters):
    guaranteed_info = """
<p>Guaranteed stock before WO is executed.</p>

<p align='left'>Calculation: 
<ul>
    <li>Current stock balance</li>
    <li>- items <b>consumed</b> by WO (at <i>start date</i>)</li>
    <li>- SO (at <i>ship date</i>)</li>
</ul>
</p>
"""

    projected_info = """
<p>Projected stock before WO is executed.</p>

<p align='left'>Calculation: 
<ul>
    <li>Current stock balance</li>
    <li>- items <b>consumed</b> by WO (at <i>start date</i>)</li>
    <li>- SO (at <i>ship date</i>)</li>
    <li>+ PO (at <i>expected delivery date</i> or <i>Reqd by date</i>)</li>
    <li>+ items <b>produced</b> by WO (at <i>expected delivery date</i> if it exists, otherwise <i>start date</i>).</li>
</ul>
</p>
"""

    cols = [
        {'fieldname': 'sufficient_stock', 'fieldtype': 'Data', 'label': _('Go?'), 'width': 50},
        {'fieldname': 'work_order', 'fieldtype': 'Data', 'label': _('Work Order'), 'options': 'Work Order', 'width': 100},
        {'fieldname': 'workstation', 'fieldtype': 'Data', 'label': _('Workstation'), 'width': 100, 'align': 'left'},
        {'fieldname': 'status', 'fieldtype': 'Data', 'label': _('Status'), 'width': 100},
        {'fieldname': 'planned_start_date', 'fieldtype': 'Date', 'label': _('Start date'), 'width': 80},
        {'fieldname': 'expected_delivery_date', 'fieldtype': 'Date', 'label': _('Delivery date'), 'width': 80},
        {'fieldname': 'qty', 'fieldtype': 'Float', 'label': _('Qty'), 'width': 100},
        {'fieldname': 'item_code', 'fieldtype': 'Data', 'label':_('Item'), 'options': 'Item', 'width': 300, 'align': 'left'},
        {'fieldname': 'sales_order', 'fieldtype': 'Link', 'label': _('Sales Order'), 'options': 'Sales Order', 'width': 100},
        {'fieldname': 'serial_no', 'fieldtype': 'Data', 'label': _('Serial No'), 'width': 200, 'align': 'left'},
        {'fieldname': 'bom_description', 'fieldtype': 'Data', 'label': _('BOM Description'), 'width': 200, 'align': 'left'},
        {'fieldname': 'comment', 'fieldtype': 'Data', 'label': _('Comment'), 'width': 200, 'align': 'left'},
    ]

    if not filters.simple_view:
        cols.extend([
            {'fieldname': 'item_group', 'fieldtype': 'Data', 'label': _('Item Group'), 'width': 100},
            {'fieldname': 'projected_stock', 'fieldtype': 'Int', 
                'label': '<span data-html="true" data-toggle="tooltip" data-placement="bottom" data-container="body" title="{}">Proj. Stock <i class="fa fa-info-circle"></i></span>'.format(projected_info), 
                'width': 110},
            {'fieldname': 'guaranteed_stock', 'fieldtype': 'Int', 
                'label': '<span data-html="true" data-toggle="tooltip" data-placement="bottom" data-container="body" title="{}">Guar. Stock <i class="fa fa-info-circle"></i></span>'.format(guaranteed_info), 
                'width': 110},
            # {'fieldname': 'projected_qty', 'fieldtype': 'Int', 'label': 'Proj. AFTER WO', 'width': 110},
            # {'fieldname': 'guaranteed_qty', 'fieldtype': 'Int', 'label': 'Guar. AFTER WO', 'width': 110},
            {'fieldname': 'stock_uom', 'fieldtype': 'Data', 'label': _('Unit'), 'width': 100},
            {'fieldname': 'warehouse', 'fieldtype': 'Data', 'label': _('Warehouse'), 'width': 100},
            {'fieldname': 'mean_time_per_unit', 'fieldtype': 'Data', 'precision': 1, 'label': _('Mean Minutes per Unit'), 'width': 100, 'align': 'right'}, # Also fomratted as data for hh:mm display
            {'fieldname': 'time_estimate_remaining', 'fieldtype': 'Data', 'label': _('Time Estimate'), 'width': 100, 'align': 'right'}, # formatted as Data for hh:mm display...
        ])

    return cols

def get_data(filters):

    # Build a full projection of the stock (similar to Projected Stock report),
    # then check for each work order the projected stock of the required items.
    # If projected stock is negative, it means there wasn't enough stock

    workstation_filter = ""
    if filters.workstation:
        workstation_filter = "AND wo2.workstation = '{}'".format(filters.workstation)

    projected_stock_query = projected_stock.build_query(so_drafts=False, wo_drafts=True, item_code=None, warehouse=None)
    sql_query = """
-- End goal: a table where each row is either a work order production item or required item, with projected stock once WO is completed.
WITH
    -- Time Estimates
    te as (SELECT
            wo.production_item AS item_code,
            AVG(wo.time_per_unit) AS mean_time_per_unit
        FROM `tabWork Order` wo
        WHERE wo.time_per_unit > 0
        GROUP BY wo.production_item)

SELECT 

    -- Common fields
    IF(p.detail_doctype = "Work Order", 0, 1) AS indent,
    p.docname AS work_order,
    p.item_code,
    ROUND(p.qty, 3) AS qty,
    -- Qty remaining AFTER WO...
    p.projected_qty,  -- ...if all planned SO, WO and PO go through
    p.guaranteed_qty, -- ...taking only items CONSUMED by WO and SO.
    -- Qty remaining BEFORE WO
    (p.projected_qty - p.qty) AS projected_stock,
    (p.guaranteed_qty - IF(p.detail_doctype = "Work Order", 0, p.qty)) AS guaranteed_stock, -- Work orders were already subtracted by a filter lower down
    p.warehouse,
    it.stock_uom,
    it.item_name,
    it.item_group,
    p.detail_doctype,
    
    -- Fields for Work Orders (produced items)
    wo.workstation,
    wo.planned_start_date,
    wo.planned_end_date,
    wo.expected_delivery_date,
    wo.qty as planned_qty,
    wo.produced_qty,
    (wo.qty - wo.produced_qty) AS required_qty,
    wo.docstatus,
    wo.status,
    wo.serial_no,
    wo.comment,
    wo.sales_order,
    wo.bom_description,
    te.mean_time_per_unit,
    te.mean_time_per_unit * (wo.qty - wo.produced_qty) AS time_estimate_remaining,

    -- Fields for Work Order Items (consumed items)
    woi.required_qty AS reqd_item_qty,
    woi.consumed_qty AS reqd_item_consumed_qty,
    (woi.required_qty - woi.consumed_qty) AS reqd_item_required_qty


FROM (
  -- This monster is the "projected stock" query: projected stock after each upcoming stock transaction (WO, SO, PO)
  {projected_stock_query}
) as p -- "projected"

LEFT JOIN `tabWork Order Item` woi on woi.name = p.detail_docname
LEFT JOIN `tabWork Order` wo on wo.name = p.detail_docname
LEFT JOIN te on wo.production_item = te.item_code
JOIN `tabItem` it on p.item_code = it.name
JOIN `tabWork Order` wo2 on p.docname = wo2.name -- To filter all rows on WO properties.

WHERE p.doctype = "Work Order"
  {workstation_filter}
ORDER BY wo2.planned_start_date, p.docname, p.detail_doctype -- Put 'Work Order Item's below 'Work Order's
    """.format(projected_stock_query=projected_stock_query, workstation_filter=workstation_filter)

    data = frappe.db.sql(sql_query, as_dict=True)

    # Check stock levels. 0 = no go, 1 = projected, 2 = guaranted
    idx = 0  # For alternating colours
    wo_go = 2
    for row in data[::-1]: 
        if row.detail_doctype == "Work Order":
            row.idx = idx
            row.sufficient_stock = wo_go
            idx += 1
            wo_go = 2
            continue
        
        go = 0
        if row.guaranteed_qty >= 0:
            go = 2
        elif row.projected_qty >= 0:
            go = 1
        row.sufficient_stock = go

        wo_go = min(go, wo_go)

    for row in data:
        row.stock_indicator = ["red", "orange", "green"][row.sufficient_stock]

    if filters.simple_view:
        data = [row for row in data if  row.docstatus == 1]

    # import pprint
    # pp = pprint.PrettyPrinter(indent=4)
    # print("\n\n\n------", pp.pprint([r for r in data if r.item_code == "100072"]))

    return data

File: bnovate/bnovate/report/work_order_planning/work_order_planning.js
// Copyright (c) 2023, bNovate, libracore and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.require("/assets/bnovate/js/lib/gcharts/loader.js")
frappe.require("/assets/bnovate/js/modals.js")  // provides bnovate.modals
frappe.require("/assets/bnovate/js/charts.js")  // provides bnovate.charts

frappe.query_reports["Work Order Planning"] = {
    filters: [
        {
            "fieldname": "workstation",
            "label": __("Workstation"),
            "fieldtype": "Link",
            "options": "Workstation"
        },
        {
            "fieldname": "simple_view",
            "label": __("Simple View"),
            "fieldtype": "Check",
        },
        {
            "fieldname": "show_chart",
            "label": __("Show Chart"),
            "fieldtype": "Check",
        },
    ],
    onload(report) {
        this.report = report;
        this.colours = ["dark", "light"];

        // report.page.add_inner_button(__('Toggle Chart'), () => {
        //     if (this.report.$chart.is(':visible')) {
        //         this.report.$chart.hide();
        //     } else {
        //         this.report.$chart.show();
        //     }
        // })

        // save a click:
        let stations = [
            'Labo',
            'Assemblage',
            'Optic',
            'Usinage',
        ];
        stations.forEach(station => report.page.add_inner_button(station, () => {
            this.report.set_filter_value('workstation', station);
        }));

        bnovate.modals.attach_report_modal("stockModal");
        bnovate.modals.attach_report_modal("cartStatusModal");
    },
    after_datatable_render(datatable) {

        if (this.report.get_filter_value('show_chart')) {
            bnovate.charts.draw_timeline_chart(this.report, build_wo_dt);
            // this.report.$chart.hide();
        }

        // Activate tooltips on columns
        $(function () {
            $('[data-toggle="tooltip"]').tooltip()
        })

    },
    formatter(value, row, col, data, default_formatter) {
        let skip_default = false;
        let simple_view = this.report.get_filter_value('simple_view');

        if (col.fieldname === "sufficient_stock") {
            const color = ['red', 'orange', 'green'][value];
            return `<span class="indicator ${color}"></span>`;
        }
        if (data.indent === 1) {
            if (['planned_start_date', 'comment', 'status'].indexOf(col.fieldname) >= 0) {
                return '';
            } else if (col.fieldname === 'item_code') {
                return projected_stock_link(data.item_code, data.warehouse, data.item_name);
            }
        } else {
            if (col.fieldname === 'comment' && value) {
                value = `<span title="${value}">${value}</span>`
            } else if (col.fieldname === 'work_order') {
                if (simple_view) {
                    value = `<a href="/desk#work-order-execution?work_order=${value}">${value}</a>`;
                } else {
                    value = frappe.utils.get_form_link("Work Order", value, true);
                }
            } else if (col.fieldname === 'serial_no' && data.serial_no) {
                value = `${cartridge_status_link(data.serial_no)}`;
            } else if (col.fieldname === 'status') {
                let [legend, colour] = work_order_indicator(data);
                return `<span class="coloured ${this.colours[data.idx % this.colours.length]}">
                        <span class="indicator ${colour}">${legend}</span>
                    </span>
                    `;
            } else if (value && (col.fieldname === 'planned_start_date' || col.fieldname === 'expected_delivery_date')) {
                let date = value.substr(0, 10);
                if (simple_view) {
                    value = date
                } else {
                    value = `<a onclick="edit_date('${data.work_order}', '${data.planned_start_date}', '${data.expected_delivery_date || ''}')">${default_formatter(date, row, col, data)}</a>`;
                    skip_default = true;
                }
            } else if (col.fieldname === 'item_code') {
                value = projected_stock_link(data.item_code, data.warehouse, data.item_name);
                skip_default = true;
            } else if (col.fieldname === 'time_estimate_remaining' && value) {
                // This will work as long as duration is less than 24h...
                value = moment.utc().startOf('day').add(value, 'minutes').format('HH:mm');
            }

            if (skip_default) {
                return `<span class="coloured ${this.colours[data.idx % this.colours.length]}">${value}</span>`;
            }
            return `<span class="coloured ${this.colours[data.idx % this.colours.length]}">${default_formatter(value, row, col, data)}</span>`;
        }
        return default_formatter(value, row, col, data)
    },
    initial_depth: 0,
};

// Stolen from work_order_list.js on ERPNext 
function work_order_indicator(doc) {
    if (doc.status === "Submitted") {
        return [__("Not Started"), "orange"];
    } else {
        return [__(doc.status), {
            "Draft": "red",
            "Stopped": "red",
            "Not Started": "red",
            "In Process": "orange",
            "Completed": "green",
            "Cancelled": "darkgrey"
        }[doc.status]];
    }
}

function cartridge_status_link(serial_nos) {
    return bnovate.modals.report_link(
        serial_nos,
        'cartStatusModal',
        'Cartridge Status',
        `Cartridge Status for this SO item`,
        {
            serial_no: serial_nos,  // Note that this is encoded in a data- tag, so it'll be a comma-separated string
        });
}

async function edit_date(work_order, start_date, delivery_date) {

    const fields = [{
        fieldname: "new_start_date",
        fieldtype: "Date",
        label: __("New Start Date"),
        default: start_date,
    }, {
        fieldname: "new_delivery_date",
        fieldtype: "Date",
        label: __("New Delivery Date"),
        default: delivery_date,
    }];

    let values = await bnovate.utils.prompt("Edit dates", fields, "Confirm", "Cancel")

    if (!values) {
        return;
    }

    await frappe.db.set_value("Work Order", work_order, "planned_start_date", values.new_start_date);
    if (values.new_delivery_date) {
        await frappe.db.set_value("Work Order", work_order, "expected_delivery_date", values.new_delivery_date);
    }
    frappe.query_report.refresh();
}

function projected_stock_link(item_code, warehouse, item_name) {
    return bnovate.modals.report_link(
        `${item_code}${item_name ? ': ' + item_name : ''}`,
        'stockModal',
        'Projected Stock',
        `Projected stock for ${item_code} in ${warehouse}`,
        {
            item_code,
            warehouse,
        });
}

function build_wo_dt(report) {
    var dataTable = new google.visualization.DataTable();

    dataTable.addColumn({ type: 'string', id: 'Workstation' });
    dataTable.addColumn({ type: 'string', id: 'Work Order' });
    dataTable.addColumn({ type: 'string', id: 'tooltip', role: 'tooltip' });
    dataTable.addColumn({ type: 'string', id: 'style', role: 'style' });
    dataTable.addColumn({ type: 'date', id: 'Start' });
    dataTable.addColumn({ type: 'date', id: 'End' });

    const frappeColors = {
        red: "#ff5858",
        orange: "#ffb65c",
        green: "#98d85b",
    }

    let rows = report.data
        .filter(row => row.indent == 0)
        .map(row => [
            row.workstation || "Unknown",
            row.work_order,
            // "fill me later",
            frappe.render_template(`
                <div class="preview-popover-header">
                    <div class="preview-header">
                    </div>

                    <div class="preview-header">
                        <div class="preview-main">
                            <a class="preview-name bold" href="#Form/Work Order/{{ work_order }}">{{ work_order }}: {{ item_name }}</a>
                            <br>
                            <span class="small preview-value">
                                Item {{ projected_stock_link(item_code, warehouse) }}
                            </span>
                            <br>
                            <span class="small text-muted">{{ planned_start_date }} {% if planned_end_date %} - {{ planned_end_date }}{% endif %}</span>
                        </div>
                    </div>
                </div>
                <hr>
                <div class="popover-body">
                    <div class="preview-table" style="max-width: none;">
                        <div class="preview-field">
                            <div class="small preview-value text-muted">
                                <table class="table table-condensed no-margin" style="border-bottom: 1px solid #d1d8dd; width:100%">
                                <thead>
                                    <th class="text-center">Qty planned</th>
                                    <th class="text-center">produced</th>
                                    <th class="text-center">remaining</th>
                                </thead>
                                <tbody>
                                    <tr>
                                        <td class="text-center">{{ planned_qty }}</td>
                                        <td class="text-center">{{ produced_qty }}</td>
                                        <td class="text-center">{{ required_qty }}</td>
                                    </tr>
                                </tbody>
                                </table>
                            </div>
                        </div>
                        <div class="preview-field">
                            <div class="small preview-value">
                                <table class="table table-condensed no-margin" style="border-bottom: 1px solid #d1d8dd; width:100%">
                                <thead>
                                    <th></th>
                                    <th class="text-left">Item</th>
                                    <th class="text-left">Name</th>
                                    <th class="text-right">Required</th>
                                    <th class="text-right">Proj. Stock</th>
                                    <th class="text-right">Guar. Stock</th>
                                </thead>
                                <tbody>
                                    {% for item in items %}
                                    <tr>
                                        <td><span class="indicator {{ ["red", "orange", "green"][item.sufficient_stock] }}"></span></td>
                                        <td class="text-left">
                                            {{ projected_stock_link(item.item_code, item.warehouse) }}
                                        </td>
                                        <td class="text-left">{{ frappe.ellipsis(item.item_name, 30) }}</td>
                                        <td class="text-right">{{ -item.qty }}</td>
                                        <td class="text-right">{{ item.projected_stock }}</td>
                                        <td class="text-right">{{ item.guaranteed_stock }}</td>
                                    </tr>
                                    {% endfor %}
                                </tbody>
                                </table>
                            </div>
                        </div>
                    </div>
                </div>
            `, {
                ...row,
                items: report.data.filter(item => item.detail_doctype === "Work Order Item" && item.work_order === row.work_order)
            }),
            `color: ${frappeColors[row.stock_indicator]}`,
            frappe.datetime.str_to_obj(row.planned_start_date),
            frappe.datetime.str_to_obj(
                row.planned_end_date ||
                row.planned_start_date + " 12:00:00"
            ),
        ])
    dataTable.addRows(rows);
    return dataTable;
}